src/libraries/System.Runtime/tests/System.Runtime.Tests/System/GCTests.cs

```diff
@@ -292,6 +293,50 @@ private class TestObject
             }
         }
 
+        // [OuterLoop]  // TODO: VS uncomment before merging
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsPreciseGcSupported))]
+        public unsafe static void WaitForPendingFinalizersRaces()
+        {
+            Task.Run(Test);
+            Task.Run(Test);
+            Task.Run(Test);
+            Task.Run(Test);
+            Task.Run(Test);
+            Task.Run(Test);
+            Test();
+
+            static void Test()
+            {
+                for (int i = 0; i < 2000; i++)
+                {
+                    bool finalized = false;
+                    MakeAndNull(&finalized);
+                    GC.Collect();
+                    GC.WaitForPendingFinalizers();
+                    Assert.True(finalized);
+                }
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            static void MakeAndNull(bool* flag)
+            {
+                var deadObj = new TestObjectWithFinalizer(flag);
+                // it's dead here
+            };
+        }
+
+        unsafe class TestObjectWithFinalizer
```

Comment:
is unsafe implementation required here?

Reply:
> is unsafe implementation required here?

Not required. I could allocate a mutable class and change a field. It is a testcase, so I used whatever was simpler. 
Accessing a local on another thread's stack is technically a UB in the memory model, so perhaps should change this to a class.